js/src/data.ts

```diff
@@ -180,8 +180,8 @@ export class NestedData<T extends NestedType = NestedType> extends BaseData<T> {
         this.childData = childData;
         this[VectorType.VALIDITY] = toTypedArray(Uint8Array, nullBitmap);
     }
-    public clone<R extends T>(type: R, length = this.length, offset = this.offset, nullCount = this._nullCount) {
-        return new NestedData<R>(type, length, this[VectorType.VALIDITY], this.childData, offset, nullCount);
+    public clone<R extends T>(type: R, length = this.length, offset = this.offset, nullCount = this._nullCount): Data<R> {
+        return new NestedData<R>(type, length, this[VectorType.VALIDITY], this.childData, offset, nullCount) as any;
```

Comment:
Does this have something to do with TS 2.9? I noticed some odd type errors when I tried building with 2.9

Reply:
Yeah it does. Seems their contravariant type checking is a bit stricter now, so have to use `any` as an escape hatch in a few places. This is one of the motivations behind slimming this class down, as I was able to remove a ton of these sorts of things.

I mentioned in slack, but can cross-post here: I've started a tentative branch that simplifies the Data down to one class ([example here](https://gist.github.com/trxcllnt/97c4332cf9d67b3eb1fcccdccadfdc0a)), and realized the two biggest hurdles is the DictionaryData and ChunkedData -- I remember why we did them, but now I'm convinced we can move their functionality to a higher level of abstraction. Going with the slim Data class brings us closer in line to the C++ implementation, and lets us remove a bunch of typings hacks along the way. Let me know your thoughts